// react-alert-with-buttons/dist/esm/index.js
import React, { createContext, useState, useEffect, useContext } from 'react';
import { createPortal } from 'react-dom';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

var AlertContext = createContext({});

var styles = {
    Background: {
        backgroundColor: "rgba(0, 0, 0, 0.5)",
        position: "fixed",
        height: "100%",
        width: "100%",
        top: 0,
        left: 0,
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        padding: "0 20px",
    },
    Container: {
        backgroundColor: "white",
        padding: 15,
        display: "flex",
        flexDirection: "column",
        width: "100%",
        maxWidth: 600,
        alignItems: "center",
    },
    CloseButton: {
        cursor: "pointer",
        alignSelf: "flex-end",
        opacity: 0.2,
        width: 40,
        height: 40,
    },
    Message: {
        alignSelf: "center",
        paddingTop: 15,
        paddingBottom: 20,
        fontSize: "1.2em",
        wordBreak: "keep-all",
        lineHeight: "1.3em",
    },
    ButtonsContainer: {
        display: "flex",
        padding: "20px 0",
        width: "100%",
        maxWidth: 460,
    },
    Button: {
        cursor: "pointer",
        color: "black",
        background: "none",
        border: "none",
        marginBottom: 10,
        padding: "15px 0",
        fontSize: "1em",
        flex: "1 0",
        backgroundColor: "#DCDCDC",
    },
};

function styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') { return; }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css_248z = "@media screen and (max-width: 576px) {\n  .AlertButtonsMediaQuery {\n    flex-direction: column;\n  }\n}\n";
styleInject(css_248z);

var Provider = function (_a) {
    var children = _a.children, containerStyle = _a.containerStyle, defaultConfirmText = _a.defaultConfirmText, buttonStyle = _a.buttonStyle;
    var _b = useState(null), rootBody = _b[0], setRootBody = _b[1];
    var _c = useState(false), isOpen = _c[0], setIsOpen = _c[1];
    var _d = useState(), alert = _d[0], setAlert = _d[1];
    var open = function (_a) {
        var message = _a.message, buttons = _a.buttons;
        setAlert({ message: message, buttons: buttons });
    };
    var close = function () {
        setIsOpen(false);
    };
    useEffect(function () {
        //open modal when open is called (open sets the message and buttons)
        if (alert) {
            setIsOpen(true);
        }
    }, [alert]);
    useEffect(function () {
        setRootBody(document.body);
    }, []);
    return (React.createElement(AlertContext.Provider, { value: { open: open, close: close } },
        rootBody &&
            isOpen &&
            alert &&
            createPortal(React.createElement("div", { style: styles.Background },
                React.createElement("div", { style: containerStyle
                        ? __assign(__assign({}, styles.Container), containerStyle) : styles.Container },
                    React.createElement("div", { style: styles.CloseButton, onClick: function () {
                            close();
                        } },
                        React.createElement("svg", { fill: "#000000", xmlns: "http://www.w3.org/2000/svg", viewBox: "0 0 50 50", width: "100%", height: "100%" },
                            React.createElement("path", { d: "M 9.15625 6.3125 L 6.3125 9.15625 L 22.15625 25 L 6.21875 40.96875 L 9.03125 43.78125 L 25 27.84375 L 40.9375 43.78125 L 43.78125 40.9375 L 27.84375 25 L 43.6875 9.15625 L 40.84375 6.3125 L 25 22.15625 Z" }))),
                    React.createElement("div", { style: styles.Message }, alert.message),
                    React.createElement("div", { style: styles.ButtonsContainer, className: "AlertButtonsMediaQuery" }, alert.buttons ? (alert.buttons.map(function (element) {
                        return (React.createElement("button", { style: element.style
                                ? __assign(__assign({}, styles.Button), element.style) : styles.Button, key: element.label, onClick: function () {
                                if (element.onClick) {
                                    element.onClick();
                                }
                            } }, element.label));
                    })) : (React.createElement("button", { style: buttonStyle
                            ? __assign(__assign({}, styles.Button), buttonStyle) : styles.Button, onClick: function () {
                            close();
                        } }, defaultConfirmText ? defaultConfirmText : "OK"))))), rootBody),
        children));
};

var useAlert = function () {
    var alertContext = useContext(AlertContext);
    return alertContext;
};

export { Provider as AlertProvider, useAlert };
